Add new section 2 about encoding choices; fix up ALDR explanations (#24)
cbor-wg/draft-ietf-cbor-cde: draft-ietf-cbor-cde.md

explicitly defined in {{-cbor}}; this document uses it in the same sense
as it is used there, specifically for applications that use CBOR as an
interchange format and use (often generic) CBOR encoders/decoders to
serialize/ingest the CBOR form of their application data to be
exchanged.

* Similarly, "CBOR Protocol" is used as in {{-cbor}} for the protocol that
governs the interchange of data in CBOR format for a specific
application or set of applications.

* "Representation" stands for the process, and its result, of building
the representation format out of (information-model level) application
data.

* "Serialization" is used for the subset of this process, and its
result, that represents ("serializes") data in CBOR generic data model
form into encoded data items.  "Encoding" is often used as a synonym
when the focus is on that.

{::boilerplate bcp14-tagged-bcp14}

# Encoding Choices in CBOR {#choi}

In many cases, CBOR provides more than one way to encode (serialize
into a sequence of bytes) a data item.
This flexibility can provide convenience for the generator of the
encoded data item, but handling the resulting variation can also put
an onus on the decoder.

> There is no one perfect encoding choice that is optimal for all

Reviewer: ```suggestion
In general, there is no single perfect encoding choice that is optimal for all
```

applications.
Choosing the right constraints on these encoding choices is one
element of application protocol design.

> Defining predefined sets of such choices is a useful way to reduce

Reviewer: ```suggestion
Having predefined sets of such choices is a useful way to reduce
```

variation between applications, enabling generic implementations.

{{Section 4.1 of RFC8949@-cbor}} provides a recommendation for a
*Preferred Serialization*.
This recommendation is useful for most CBOR applications, and it is a
good choice for most applications.
Its main constraint is to choose the shortest _head_ ({{Section 3
of RFC8949@-cbor}}) that preserves the value of a data item.

Preferred Serialization allows indefinite length encoding ({{Section
3.2 of RFC8949@-cbor}}), which does not express the length of a string,
an array, or a map in its head.  Supporting both definite length and
indefinite length encoding is an additional onus on the decoder; many
applications therefore choose not to use indefinite length encoding at
all.
We call Preferred Serialization with this additional constraint
*Basic Serialization*.
Basic Serialization is a common choice for applications that need to
further reduce the variability that needs to be handled by decoders,
potentially maximizing interoperability with partial (e.g.,
constrained) CBOR decoder implementations.

> These constraints still allow some variation, in particular, there is

Reviewer: ```suggestion
These constraints still allow some variation. In particular, there is
```

more than one serialization for data items that contain maps: The
order of serialization of map entries is ignored in CBOR (as it is in
JSON), so maps with more than one entry have all permutations of these
entries as valid Basic Serializations.
*Deterministic Serialization* builds on Basic Serialization by

> defining a common (namely, lexicographic) order for the entries in a map.

Reviewer: ```suggestion
defining a deterministic (namely, lexicographic) order for the keys in a map.
```

Author: Hmm, you don't order the keys, you order the entries (by their keys).
Also, I tried to point out that we need to agree on an order (so that in the end gives a deterministic result, of course, but the agreement was the point).

For many applications, ensuring this common order is an additional
onus on the generator that is not actually needed, so they do not
choose Deterministic Serialization.
However, if the objective is minimal effort for the consuming
application, deterministic map ordering can be useful even outside the
main use cases for Deterministic Serialization that are further
discussed in {{Section 2 of -det}}.

{{tab-constraints}} summarizes the increasingly restrictive sets of
encoding choices that have been given names in this section.

{: #tab-constraints title="Constraints on the Serialization of CBOR"}
| Set of Encoding Choices | Most Important Constraint | Applications |
|-------------------------|---------------------------|--------------|
| preferred               | shortest "head" variant   | most         |
| basic                   | + definite lengths only   | many         |

> | *deterministic* ("CDE") | + common map order        | specific     |

Reviewer: I like!

Note that the objective to have a deterministic serialization for a
specific application data item can only be fulfilled if the
application itself does not generate multiple different CBOR data

> items that represent that same (equivalent) application data item.

Reviewer: I'm not sure I can parse this. How would an application "generate (...) CBOR data items"? Shouldn't applications be concerned with application data items and not CBOR data items?

I guess you want to allude to problems such as integer/float equivalence on the application layer, but I can't match that to the sentence as is.

Isn't it then more about the application handling different (received) CBOR data items as the same application data item?

Author: Well, something needs to go from the application's information model to the CBOR generic data model.  Generally, this needs to be part of the application -- generally not the business end, but somewhere on the belly.  The onus on being deterministic is on the originating application.  To deal with bugs and attacks, the ingesting application should then check that this is done properly.

Reviewer: Okay, makes sense. Thanks for re-explaining!

We speak of the need for Application-level Deterministic
Representation (ALDR), and we may want to aid achieving this by

> the application defining rules for ALDR (see also {#aldr}).

Reviewer: ```suggestion
the application defining rules for ALDR (see also {{aldr}}).
```
(not sure about the right hyperlink syntax)

Where Deterministic Representation is not actually needed,

> Application-level representation rules of course can still be useful

Reviewer: ```suggestion
application-level representation rules of course can still be useful
```

> to amplify the benefits of Preferred or Basic Serialization.

Reviewer: ```suggestion
to amplify the benefits of Preferred or Basic Serialization,
but would not fall into the category of ALDR rules.
```

Couldn't we instead loosen the term ALDR (to ALR?) to encompass those, too? The differentiation between CBOR layer encoding rules and application-layer encoding rules would also be present for those cases, and would benefit from a term, too, I'd say.

Author: The representation of application data in CBOR data items generally is governed by rules, which we could indeed term ALRs.  But this is so obvious that there is little point do have a term for that.

The need for ALDRs, in contrast, is specific to deterministic representation, and it can be surprising if you haven't done deterministic representation of application data before.
So it is a good thing to point this out and to also relieve the reader from the potental misunderstanding that ALDR needs to be put into CDE, because all deterministic encoding is put there.

Reviewer: ACK :)
